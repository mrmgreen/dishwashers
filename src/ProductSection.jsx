import React from 'react';

function ProductSection(props) {
  const { title, price: { now }, image } = props.product;
  let priceToPay;

  if (typeof now === "object") {
    priceToPay = now.to;
  }

  return ( 
    <div className="productSection">
      <div className="product_section_top">
        <img src={image} alt="dishwasher"/>
      </div>
      <div className="product_section_bottom">
        <div className='product_title'>{title}</div>
        <div className='price'>Â£{ priceToPay || now}</div>
      </div>
    </div>
  )
}

export default ProductSection;