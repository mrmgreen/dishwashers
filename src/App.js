import React, { Component } from 'react';
import axios from 'axios';
import './App.css';
import filterProductsGrid from './utils/filterProductsGrid';
import ProductSection from './ProductSection';

class App extends Component {
  _isMounted = false;

  constructor(props) {
    super(props);
    this.state = {
      products: []
    }
  }

  componentDidMount() {
    this._isMounted = true;

    // const gridAPI = 'https://api.johnlewis.com/v1/products/search?q=dishwasher&key=Wu1Xqn3vNrd1p7hqkvB6hEu0G9OrsYGb&pageSize=20';
    const gridAPI = 'http://localhost:4000/grid';
    axios.get(gridAPI)
      .then(response => {
        if (response.data && response.data.products !== undefined ) {
          const products = filterProductsGrid(response.data.products);
          if (this._isMounted){
            this.setState({ products });
          }
        }
      })
      .catch(e => console.log('Error fetching Grid API', e.message))
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
        { this.state.products.length > 0 &&
            this.state.products.map((product, id) => <ProductSection key={id} product={product} />)
        }
        </header>
      </div>
    );
  }
}

export default App;
