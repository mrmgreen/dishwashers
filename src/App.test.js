import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from './App';
import axios from 'axios';
import ProductSection from './ProductSection';

jest.mock('axios');
const productsData = [
  {
    productId: 1391191,
    title:"Indesit DIF 04B1 Ecotime Fully Integrated Dishwasher, White",
    price:{  
      was:"",
      then1:"",
      then2:"",
      now:"220.00",
      uom:"",
      currency:"GBP"
    },
    image: "//johnlewis.scene7.com/is/image/JohnLewis/233326789?",
  }
];
// const gridAPI = 'https://api.johnlewis.com/v1/products/search?q=dishwasher&key=Wu1Xqn3vNrd1p7hqkvB6hEu0G9OrsYGb&pageSize=20';
const gridAPI = 'http://localhost:4000/grid';
axios.get.mockResolvedValue({data: { products: productsData }});
let component;

beforeEach(() => {
  axios.get.mockClear();
  component = shallow(<App />);
});

it('renders without crashing', () => {
  expect(component).toHaveLength(1);
});

it('makes a request to the product grid API when the grid page loads', () => {
  expect(axios.get).toBeCalledTimes(1);
  expect(axios.get).toBeCalledWith(gridAPI);
});

it('should render a ProductSection for each product', () => {
  expect(component.find(ProductSection)).toHaveLength(1);
});

it('render page with products', () => {
  expect(component).toMatchSnapshot();
});