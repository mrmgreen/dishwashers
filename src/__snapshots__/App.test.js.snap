// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render page with products 1`] = `
<div
  className="app"
>
  <header>
    Dishwashers
  </header>
  <div
    className="products__container"
  >
    <ProductSection
      key="0"
      product={
        Object {
          "image": "//johnlewis.scene7.com/is/image/JohnLewis/233326789?",
          "price": Object {
            "currency": "GBP",
            "now": "220.00",
            "then1": "",
            "then2": "",
            "uom": "",
            "was": "",
          },
          "productId": 1391191,
          "title": "Indesit DIF 04B1 Ecotime Fully Integrated Dishwasher, White",
        }
      }
    />
  </div>
</div>
`;
